src/modules/entities/user/services/user.service.ts

```diff
@@ -0,0 +1,150 @@
+    @InjectRepository(User)
+    private readonly userRepository: Repository<User>,
+    private readonly bcrypt: Bcrypt,
+  ) {}
+
+  async create(createUserDto: CreateUserDto) {
+    await Promise.all([
+      this.checkDataExist('email', createUserDto.email),
+      this.checkDataExist('phone_number', createUserDto.phone_number),
+    ]);
+
+    const { password_hash, password_salt } = await this.hashPassword(
+      createUserDto.password,
+    );
+
+    const user = await this.userRepository.save({
+      password_hash,
+      password_salt,
+      ...createUserDto,
+    });
+
+    delete user.password_hash;
+    delete user.password_salt;
+
+    return user;
+  }
+
+  async findAll() {
+    return await this.userRepository.find({});
+  }
+
+  async findOneById(id: number) {
+    return await this.userRepository.findOne({ where: { id } });
+  }
+
+  async findOneByMail(email: string) {
+    return await this.userRepository.findOne({ where: { email } });
+  }
+
+  async findOneByPhone(phone: string) {
+    return await this.userRepository.findOne({
+      where: { phone_number: phone },
+    });
+  }
+
+  async findOneByUsername(username: string) {
+    return await this.userRepository.find({
+      where: { first_name: username },
+    });
+  }
+
+  async update(id: number, updateUser: UpdateUserDto) {
+    await this.checkDataExist('id', id.toString(), id);
+    if (updateUser.email)
+      await this.checkDataExist('email', updateUser.email, id);
+    if (updateUser.phone_number)
+      await this.checkDataExist('phone_number', updateUser.phone_number, id);
+
+    await this.userRepository.update(id, updateUser);
+    return await this.userRepository.findOne({ where: { id } });
+  }
+
+  async remove(id: number) {
+    await this.checkDataExist('id', id.toString());
+    return await this.userRepository.delete(id);
+  }
+
+  async updatePassword(
+    id: number,
+    currentPassword: string,
+    newPassword: string,
+  ) {
+    await this.confirmPassword(id, currentPassword);
+    const { password_hash, password_salt } =
+      await this.hashPassword(newPassword);
+
+    await this.userRepository.update(id, {
+      password_hash,
+      password_salt,
+    });
+
+    return {
+      message: 'Password updated',
+    };
+  }
+
+  async confirmPassword(id: number, password: string) {
+    const user = await this.userRepository
+      .createQueryBuilder('user')
+      .where('id = :id', { id })
+      .addSelect('user.password_hash')
+      .getOne();
+
+    const isValidPassword = await this.bcrypt.compare(
+      password,
+      user.password_hash,
+    );
+
+    if (!isValidPassword)
+      throw new HttpException(
+        'Current password is wrong',
+        HttpStatus.BAD_REQUEST,
+      );
+  }
+
+  async hashPassword(password: string) {
+    const password_salt = await this.bcrypt.getSalt();
+    const password_hash = await this.bcrypt.hash(password, password_salt);
+    return { password_hash, password_salt };
+  }
```

Comment:
for this is usually better to create a separatr service that is taking care of security like AuthenticationService than mixing with the general User logic

Reply:
I managed to put the "hashPassword" method in the bcrypt service. But I got some circular dependency errors when I tried to put the "confirmPassword" method in the authentication service.